presentation-api/controlling-ua/PresentationConnection_onterminated-manual.html

```diff
@@ -21,34 +21,40 @@
 <script>
   setup({explicit_timeout: true});
   var startPresentation = function () {
+
     document.getElementById('presentBtn').disabled = true;
-    async_test(function(t) {
+
+    promise_test(function (t) {
       var request = new PresentationRequest(presentationUrls);
-      request.start()
-        .then(function(connection) {
+
+      return request.start()
+        .then(function (connection) {
 
           // Enable timeout again, cause no user action is needed from here.
-          t.step_timeout(function() {
+          t.step_timeout(function () {
               t.force_timeout();
               t.done();
           }, 5000);
 
           assert_true(connection instanceof PresentationConnection);
-          connection.onconnect = t.step_func(function(evt) {
+
+          connection.onconnect = function (evt) {
             assert_equals(connection.state, "connected");
             connection.terminate();
-          });
-          connection.onterminate = t.step_func_done(function(evt) {
+          };
+
+          connection.onterminate = function (evt) {
             assert_equals(evt.type, "terminate");
             assert_equals(connection.state, "terminated");
-          });
-          connection.onclose = t.step_func_done(function(evt) {
+          };
+
+          connection.onclose = function (evt) {
```

Comment:
Right. Which should boil down to something like:

```js
let connection = null;
return request.start()
  .then(function (c) {
    connection = c;
    [...]
    const watcher = new EventWatcher(t, connection, 'connect');
    return watcher.wait_for('connect');
  })
  .then(function () {
    connection.terminate();
    const watcher = new EventWatcher(t, connection, ['terminate', 'close']);
    return watcher.wait_for('terminate');
  })
  .then(function () {
    assert_equals(connection.state, 'terminated');
  });
```

or you could perhaps use the same `EventWatcher` as in:

```js
let connection = null;
let watcher = null;
return request.start()
  .then(function (c) {
    connection = c;
    [...]
    watcher = new EventWatcher(t, connection, ['connect', 'close', 'terminate']);
    return watcher.wait_for('connect');
  })
  .then(function () {
    connection.terminate();
    return watcher.wait_for('terminate');
  })
  .then(function () {
    assert_equals(connection.state, 'terminated');
  });
```